ibis/backends/base/__init__.py

```diff
@@ -903,22 +914,40 @@ def _cache(self, expr):
             Cached table
 
         """
-        raise NotImplementedError(
-            f"{self.name} backend has not implemented `cache` API"
-        )
-
-    def _release_cache(self, expr):
+        op = expr.op()
+        if (result := self._cache.get(op)) is None:
+            name = util.generate_unique_table_name("cache")
+            self._load_into_cache(name, expr)
+            self._cache[op] = result = self.table(name, expr.schema()).op()
+        self._refs[op] += 1
```

Comment:
Could we factor this out into a refcounted cache container under `ibis.common.caching` (e.g. using callbacks)?

I'd prefer to see this as a standalone utility because:
1. It's cleaner to test the refcounting, unique-key generation, and cache operations
2. It would be useful for other purposes, e.g. as a result cache for the pandas backend
3. Harder to make mistakes if we refactor the backend class

Reply:
Roger. I've started this in a branch called [`ref-counted-cache`](https://github.com/cpcloud/ibis/tree/ref-counted-cache) that I will PR after this one lands!